From: Simon McVittie <smcv@collabora.com>
Date: Thu, 20 May 2021 18:52:38 +0100
Subject: portal: Don't leak fd used for serialized environment

Otherwise we'll run out of file descriptors eventually, when starting
a sufficiently large number of subsandboxes.

Resolves: flatpak/flatpak#4285
Fixes: aeb6a7ab "portal: Convert --env in extra-args into --env-fd"
Signed-off-by: Simon McVittie <smcv@collabora.com>
Origin: upstream, 1.11.2, commit:f2fbc75827a58cc6b4cba48a0c895c3313274020
Applied-upstream: 1.10.3, commit:b4c6aa1cc88625db8b5a26313b34bad3e7c2ed56
Forwarded: https://github.com/flatpak/flatpak/pull/4322
---
 portal/flatpak-portal.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/portal/flatpak-portal.c b/portal/flatpak-portal.c
index 0539ff2..e6eeecb 100644
--- a/portal/flatpak-portal.c
+++ b/portal/flatpak-portal.c
@@ -788,6 +788,7 @@ handle_spawn (PortalFlatpak         *object,
   gboolean notify_start;
   gboolean devel;
   g_autoptr(GString) env_string = g_string_new ("");
+  glnx_autofd int env_fd = -1;
 
   child_setup_data.instance_id_fd = -1;
   child_setup_data.env_fd = -1;
@@ -1122,10 +1123,10 @@ handle_spawn (PortalFlatpak         *object,
           return G_DBUS_METHOD_INVOCATION_HANDLED;
         }
 
-      child_setup_data.env_fd = glnx_steal_fd (&env_tmpf.fd);
+      env_fd = glnx_steal_fd (&env_tmpf.fd);
+      child_setup_data.env_fd = env_fd;
       g_ptr_array_add (flatpak_argv,
-                       g_strdup_printf ("--env-fd=%d",
-                                        child_setup_data.env_fd));
+                       g_strdup_printf ("--env-fd=%d", env_fd));
     }
 
   for (i = 0; unset_env != NULL && unset_env[i] != NULL; i++)
